services/yellow/src/api/Handlers/AddPaymentMethodToOrderHandler.cs

```diff
@@ -20,30 +22,32 @@ public AddPaymentMethodToOrderHandler(IOrderService orderService, ICreditCardPay
             _paymentService = paymentService;
         }
 
-        protected async override Task<PaymentMethodToOrderAdded> Handle(AddPaymentMethodToOrderRequest message)
+        public async Task Handle(AddPaymentMethodToOrderRequest message, IMessageHandlerContext context)
         {
-            var order = await Ensure(message, m => _orderService.FindOneById(m.OrderId));
-
-            var paymentToken = await _paymentService.GetPaymentToken(message.PaymentMethod.CreditCard!);
-            var orderPaymentMethod = new StoredPaymentMethod(message.PaymentMethod.CreditCard!.GetOrderCreditCard(paymentToken));
+            try
+            {
+                var validateResult = ValidateRequest(message);
+                if (!validateResult.IsValid())
+                    throw new ValidationException(validateResult);
             
-            await _orderService.AddPaymentMethod(order, orderPaymentMethod);
-
-            return new PaymentMethodToOrderAdded(message.OrderId);
-        }
-        
-        protected override object Fail(AddPaymentMethodToOrderRequest requestMessage)
-        {
-            return new AddPaymentMethodToOrderFailed(requestMessage.OrderId);
-        }
-
-        protected override Task AddPaymentMethod(Order order, StoredPaymentMethod paymentMethod)
-        {
-            return _orderService.AddPaymentMethod(order, paymentMethod);
+                var order = HandlerHelper.Ensure(await _orderService.FindOneById(message.OrderId));
+            
+                var paymentToken = await _paymentService.GetPaymentToken(message.PaymentMethod.CreditCard!);
+                var orderPaymentMethod = new StoredPaymentMethod(message.PaymentMethod.CreditCard!.GetOrderCreditCard(paymentToken));
+            
+                await _orderService.AddPaymentMethod(order, orderPaymentMethod);
+            
+                await context.Publish(new PaymentMethodToOrderAdded(message.OrderId));
+                await context.Reply(new Response<bool>(true));
+            }
+            catch (Exception e)
```

Comment:
If we catch everything here, we loose the benefits of asynchronous messaging.  If we only catch the `ValidationException` here, it would make sense to the client. But if something else fails, I want the system to try again (message gets requeued if there is an exception) and maybe sent to an error queue.

Reply:
suggestion applied!